### PR TITLE
Update the documentation

### PR DESCRIPTION
**Pull request by djs55**
- add a Windows and Mac example plumbing diagram
- add `SOCK_STREAM` and `SOCK_DGRAM` to the main `vpnkit` diagram
- link to hyperkit and Mirage TCP/IP
- add a high-level walk-through describing what happens when a process in a Linux VM connected to `vpnkit` attempts to establish a TCP connection

### CODE REVIEW

File: docs/ethernet.md [hunk 1]
```diff
@@ -1,28 +1,118 @@
...
-- if they contain IPv4 datagrams then we create a fresh TCP/IP stack, a fresh
-  switch port and connect them together so that all future IPv4 traffic to the
-  same destination address is processed by the new stack.
+  table
+- if they contain IPv4 datagrams then we create a fresh virtual TCP/IP endpoint
+  using the [Mirage](https://mirage.io/) [TCP/IP stack](https://github.com/mirage/mirage-tcpip)
+  (no kernel TCP/IP interfaces are involved), a fresh
+  switch port on our internal switch and connect them together so that all future IPv4 traffic to the
+  same destination address is processed by the new endpoint.
+
```
Comment (MagnusS) [inline docs/ethernet.md]: There's also the L2 bridge in there now - so frames can come from any other Vpnkit instance via the bridge or from the connected VM: https://github.com/moby/vpnkit/blob/master/src/hostnet/slirp.ml#L731
 -- and frames from the VM without server/client/broadcast dst addresses will be sent to the bridge https://github.com/moby/vpnkit/blob/master/src/hostnet/slirp.ml#L749

The static ARP table also includes mac addresses to other Vpnkit instances.

Comment (djs55) [inline docs/ethernet.md]: Cool -- are you volunteering to write a little bit about it? :) Perhaps I'll merge this and then you could make a PR on top?

Comment (MagnusS) [inline docs/ethernet.md]: Sure, I can add more details after this is merged

File: docs/ethernet.md [hunk 2]
```diff
@@ -1,28 +1,118 @@
...
+Note that from the host kernel's point of view, there is no network connection
+to the VM and no set of associated firewall rules or routing tables. All outgoing
+connections originate from the `vpnkit` process. If the user installs some
+advanced networking or VPN software which reconfigures the routing table or
+firewall rules, it will not break the connection between `vpnkit` and the VM.
 
-The active ports may be queried over the 9P debug interface via the `ports`
-subdirectory.
+This technique for forwarding network connections is commonly known as
+[Slirp](https://en.wikipedia.org/wiki/Slirp).
```
Comment (samoht) [inline docs/ethernet.md]: The wikipedia page says:
>  Even though the Slirp-derived code was heavily criticized,[11] to the date there is no competing implementation available.

Maybe someone can try to update that page? :p /cc @amirmc

**Pull request merged**